### ISSUE TITLE
RuntimeError: 60 elements of the 60 element gradient array `gradf` are NaN. This often indicates numerical issues.

### ISSUE THREAD

**Issue by FelixNeutatz** [facebook/Ax#576]
Hi everybody,

I run into the following nondeterministic exception:
```
batch = exp.new_trial(generator_run=gpei.gen(1))
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/modelbridge/base.py", line 628, in gen
    observation_features, weights, best_obsf, gen_metadata = self._gen(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/modelbridge/array.py", line 250, in _gen
    X, w, gen_metadata, candidate_metadata = self._model_gen(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/modelbridge/torch.py", line 219, in _model_gen
    X, w, gen_metadata, candidate_metadata = self.model.gen(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/torch/botorch.py", line 390, in gen
    candidates, expected_acquisition_value = make_and_optimize_acqf()
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/torch/botorch.py", line 376, in make_and_optimize_acqf
    candidates, expected_acquisition_value = self.acqf_optimizer(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/torch/botorch_defaults.py", line 297, in scipy_optimizer
    X, expected_acquisition_value = optimize_acqf(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/botorch/optim/optimize.py", line 166, in optimize_acqf
    batch_candidates_curr, batch_acq_values_curr = gen_candidates_scipy(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/botorch/generation/gen.py", line 129, in gen_candidates_scipy
    res = minimize(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/scipy/optimize/_minimize.py", line 617, in minimize
    return _minimize_slsqp(fun, x0, args, jac, bounds,
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/scipy/optimize/slsqp.py", line 399, in _minimize_slsqp
    fx = func(x)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/scipy/optimize/optimize.py", line 327, in function_wrapper
    return function(*(wrapper_args + args))
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/scipy/optimize/optimize.py", line 65, in __call__
    fg = self.fun(x, *args)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/botorch/generation/gen.py", line 125, in f
    raise RuntimeError(msg)
RuntimeError: 60 elements of the 60 element gradient array `gradf` are NaN. This often indicates numerical issues.
```
with the following program:
```
from ax import *
space = SearchSpace(parameters=[
    ChoiceParameter('class_weighting', parameter_type=ParameterType.BOOL, values=[True, False], is_ordered=False),
    ChoiceParameter('custom_weighting', parameter_type=ParameterType.BOOL, values=[True, False], is_ordered=False),
    RangeParameter(name="custom_weight", parameter_type=ParameterType.FLOAT, lower=0.00000001, upper=1.0)])
  
exp = SimpleExperiment(
            name="test",
            search_space=space,
            evaluation_function=my_objective,
            objective_name='accuracy',
            minimize=False
        )

sobol = Models.SOBOL(exp.search_space)
for i in range(5):
     exp.new_trial(generator_run=sobol.gen(1))

for i in range(25):
     gpei = Models.BOTORCH(experiment=exp, data=exp.eval())
     batch = exp.new_trial(generator_run=gpei.gen(1))
```

If I run it with Models.GPEI instead, I get the following exception
```
Traceback (most recent call last):
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/utils/cholesky.py", line 32, in psd_safe_cholesky
    L = torch.cholesky(A, upper=upper, out=out)
RuntimeError: cholesky_cpu: For batch 3: U(7,7) is zero, singular U.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
      batch = exp.new_trial(generator_run=gpei.gen(1))
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/modelbridge/base.py", line 628, in gen
    observation_features, weights, best_obsf, gen_metadata = self._gen(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/modelbridge/array.py", line 267, in _gen
    xbest = self._model_best_point(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/modelbridge/torch.py", line 259, in _model_best_point
    X = self.model.best_point(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/torch/botorch.py", line 419, in best_point
    return self.best_point_recommender(  # pyre-ignore [28]
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/torch/botorch_defaults.py", line 354, in recommend_best_observed_point
    x_best = best_observed_point(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/model_utils.py", line 290, in best_observed_point
    best_point_and_value = best_in_sample_point(
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/model_utils.py", line 401, in best_in_sample_point
    f, cov = as_array(model.predict(X_obs))
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/torch/botorch.py", line 306, in predict
    return self.model_predictor(model=self.model, X=X)  # pyre-ignore [28]
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/ax/models/torch/utils.py", line 463, in predict_from_model
    posterior = model.posterior(X)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/botorch/models/gpytorch.py", line 325, in posterior
    mvn = self(X)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/models/exact_gp.py", line 319, in __call__
    predictive_mean, predictive_covar = self.prediction_strategy.exact_prediction(full_mean, full_covar)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/models/exact_prediction_strategies.py", line 262, in exact_prediction
    self.exact_predictive_mean(test_mean, test_train_covar),
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/models/exact_prediction_strategies.py", line 280, in exact_predictive_mean
    res = (test_train_covar @ self.mean_cache.unsqueeze(-1)).squeeze(-1)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/utils/memoize.py", line 59, in g
    return _add_to_cache(self, cache_name, method(self, *args, **kwargs), *args, kwargs_pkl=kwargs_pkl)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/models/exact_prediction_strategies.py", line 229, in mean_cache
    mean_cache = train_train_covar.evaluate_kernel().inv_matmul(train_labels_offset).squeeze(-1)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py", line 1180, in inv_matmul
    return func.apply(self.representation_tree(), False, right_tensor, *self.representation())
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/functions/_inv_matmul.py", line 53, in forward
    solves = _solve(lazy_tsr, right_tensor)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/functions/_inv_matmul.py", line 17, in _solve
    return lazy_tsr.cholesky()._cholesky_solve(rhs)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py", line 967, in cholesky
    chol = self._cholesky(upper=False)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/utils/memoize.py", line 59, in g
    return _add_to_cache(self, cache_name, method(self, *args, **kwargs), *args, kwargs_pkl=kwargs_pkl)
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py", line 429, in _cholesky
    cholesky = psd_safe_cholesky(evaluated_mat, upper=upper).contiguous()
  File "/root/anaconda3/envs/DeclarativeAutoML/lib/python3.8/site-packages/gpytorch/utils/cholesky.py", line 55, in psd_safe_cholesky
    raise NotPSDError(
gpytorch.utils.errors.NotPSDError: Matrix not positive definite after repeatedly adding jitter up to 1.0e-06. Original error on first attempt: cholesky_cpu: For batch 3: U(7,7) is zero, singular U.

```

I appreciate any hints to get it to work :)

Best regards,
Felix

**Comment by Jakepodell** [comment]
Hey Felix,

I'm taking a look and we will have a response shortly!

**Comment by lena-kashtelyan** [comment]
Hi @FelixNeutatz, thank you for reporting this issue! I think most likely this is a duplicate of https://github.com/facebook/Ax/issues/228, and the Cholesky errors you are seeing indicate that repeated trials are being suggested and the optimization converged, so you can safely consider it concluded.

`Models.GPEI` and `Models.BOTORCH` are actually the same under the hood currently, so I think this error is just somewhat non-deterministic. 

cc @qingfeng10 as the modeling and optimization oncall and @Balandat for additional thoughts

**Comment by FelixNeutatz** [comment]
Hi @lena-kashtelyan,

thank you for the quick answer. I checked the configurations and it was exactly as you describe in #228, the optimizer chose similar configurations over and over. As this hints convergence, we can assume that once a exception appears, our optimization is done, right? Therfore, I just do this:
```
sobol = Models.SOBOL(exp.search_space)
for i in range(5):
     exp.new_trial(generator_run=sobol.gen(1))

for i in range(25):
     try:
         gpei = Models.BOTORCH((experiment=exp, data=exp.eval())
         batch = exp.new_trial(generator_run=gpei.gen(1))
     except:
         break
```

**Comment by lena-kashtelyan** [comment]
Yep, I think you can do exactly that! We will try and exit more gracefully in the long-term, but this has been a difficult issue to solve. We are working towards solving it with methodological improvements rather than just exiting on these errors, since they could indicate numerical issues and not just convergence. However, in the case of repeated suggestions convergence is quite safe to assume.
